include/envoy/router/router.h

```diff
@@ -876,11 +896,12 @@ class Config {
    * @param headers supplies the request headers.
    * @param random_value supplies the random seed to use if a runtime choice is required. This
    *        allows stable choices between calls if desired.
+   * @param route_index specifies which route index to start looking for route on.
```

Comment:
I'm not sure that the index is going to be relevant if/when the route table contains hash lookup types.

Stepping back, I'm sorry but I continue to have strong reservations that this is a general purpose feature that is useful to people. The invasiveness of this change makes me think we should look for a better way that you could handle as an extension.

What I'm thinking is:
1) What if we allows the route function to take an optional callback that asks that caller to decide if a route should be accepted?
2) You can store fall through status inside route metadata.
3) Make it so that you can write a filter that can repopulate the route using a callback function to select the final route, and do whatever logging you want along the way.

^ seems much less invasive while still allowing you to achieve your goals. @htuch WDYT?

Reply:
@mattklein123 are you describing something similar to the host retry picking extension approach but for routes? I kind of like where this is heading, in that we are making routing more extensible.

@alyssawilk and I have talked in the past about making routing pluggable in general; one of the challenges with a pluggable router today is that it forces you to lose the HCM/router routing logic and build it completely from scratch. I wonder if what you propose, or some variant, could be the solution here. Also CC @markdroth, we've been discussing the possibility of having a way to say "the cluster for this route comes from an extension" in an unrelated conversation.

Reply:
> @mattklein123 are you describing something similar to the host retry picking extension approach but for routes? I kind of like where this is heading, in that we are making routing more extensible.

Yeah pretty much. I think we could do this pretty cleanly without even a new extension point (just an HTTP filter perhaps, though depending on where we want the route to be picked I suppose it might need to be a first-class extension that is runnable directly from the HCM code).

Reply:
@htuch and @mattklein123 if you both agree to the approach which @mattklein123 described above, then i can start making changes.

What if we allows the route function to take an optional callback that asks that caller to decide if a route should be accepted?
You can store fall through status inside route metadata.
Make it so that you can write a filter that can repopulate the route using a callback function to select the final route, and do whatever logging you want along the way.

Reply:
I think this approach will work for me. @htuch ?

Reply:
Yeah, this sounds the right direction. @pitiwari can you put something up for review when you have the basic interfaces worked out, before going too far into writing tests etc? I'd like to see see what this looks like in code.

Reply:
@htuch sure i will do it